### ISSUE TITLE
LXC: Segmentation fault, sporadically

### ISSUE THREAD

**Issue by Maximus43** [ClickHouse/ClickHouse#21594]
Sometimes, once in 2-3 hours, Clickhouse restarts with segmentation fault. Database has almost no load, approx. 1000 inserts per minute.

**Does it reproduce on recent release?**
Yes

**How to reproduce**
I don't know, it happens sporadically

* Which ClickHouse server version to use
21.2.5.5

* Which interface to use, if matters
ClickHouse Python Driver (clickhouse-driver 0.2.0)

**Expected behavior**
No segfaults

**Error message and/or stacktrace**
```
2021.03.10 13:36:18.523605 [ 19126 ] {} <Fatal> BaseDaemon: ########################################
2021.03.10 13:36:18.523690 [ 19126 ] {} <Fatal> BaseDaemon: (version 21.2.5.5 (official build), build id: BFF9D4E9C079763C6DFA1C04E8A4D1903D0FFE69) (from thread 19020) (no query) Received signal Segmentation fault (11)
2021.03.10 13:36:18.523713 [ 19126 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2021.03.10 13:36:18.523732 [ 19126 ] {} <Fatal> BaseDaemon: Stack trace: 0xe85d688 0xe859455 0xe85c5a8 0xe85d152 0x84f700f 0x84faaa3 0x7fac958dd4a4 0x7fac9561fd0f
2021.03.10 13:36:18.523811 [ 19126 ] {} <Fatal> BaseDaemon: 2. std::__1::__tree<std::__1::__value_type<Poco::Timestamp, std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> >, std::__1::__map_value_compare<Poco::Timestamp, std::__1::__value_type<Poco::Timestamp, std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> >, std::__1::less<Poco::Timestamp>, true>, std::__1::allocator<std::__1::__value_type<Poco::Timestamp, std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> > > >::erase(std::__1::__tree_const_iterator<std::__1::__value_type<Poco::Timestamp, std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> >, std::__1::__tree_node<std::__1::__value_type<Poco::Timestamp, std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> >, void*>*, long>) @ 0xe85d688 in /usr/bin/clickhouse
2021.03.10 13:36:18.523853 [ 19126 ] {} <Fatal> BaseDaemon: 3. DB::BackgroundSchedulePoolTaskInfo::scheduleImpl(std::__1::lock_guard<std::__1::mutex>&) @ 0xe859455 in /usr/bin/clickhouse
2021.03.10 13:36:18.523871 [ 19126 ] {} <Fatal> BaseDaemon: 4. DB::BackgroundSchedulePool::delayExecutionThreadFunction() @ 0xe85c5a8 in /usr/bin/clickhouse
2021.03.10 13:36:18.523886 [ 19126 ] {} <Fatal> BaseDaemon: 5. ? @ 0xe85d152 in /usr/bin/clickhouse
2021.03.10 13:36:18.523904 [ 19126 ] {} <Fatal> BaseDaemon: 6. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x84f700f in /usr/bin/clickhouse
2021.03.10 13:36:18.523919 [ 19126 ] {} <Fatal> BaseDaemon: 7. ? @ 0x84faaa3 in /usr/bin/clickhouse
2021.03.10 13:36:18.523942 [ 19126 ] {} <Fatal> BaseDaemon: 8. start_thread @ 0x74a4 in /lib/x86_64-linux-gnu/libpthread-2.24.so
2021.03.10 13:36:18.523963 [ 19126 ] {} <Fatal> BaseDaemon: 9. clone @ 0xe8d0f in /lib/x86_64-linux-gnu/libc-2.24.so
2021.03.10 13:36:18.615505 [ 19126 ] {} <Fatal> BaseDaemon: Checksum of the binary: 94B88F83D23FC43FAEFCE896EDC6A812, integrity check passed.
2021.03.10 13:36:28.554129 [ 18967 ] {} <Fatal> Application: Child process was terminated by signal 11.
2021.03.10 13:36:58.997144 [ 19522 ] {} <Error> bool DB::(anonymous namespace)::checkPermissionsImpl(): Code: 412, e.displayText() = DB::Exception: Can't receive Netlink response: error -2, Stack trace (when copying this message, always include the lines below):

0. ? @ 0x84eadb2 in /usr/bin/clickhouse
1. ? @ 0x84eaf08 in /usr/bin/clickhouse
2. DB::TaskStatsInfoGetter::TaskStatsInfoGetter() @ 0x84ea837 in /usr/bin/clickhouse
3. ? @ 0x84ea686 in /usr/bin/clickhouse
4. DB::TaskStatsInfoGetter::checkPermissions() @ 0x84ea626 in /usr/bin/clickhouse
5. DB::TasksStatsCounters::create(unsigned long) @ 0x84e2e00 in /usr/bin/clickhouse
6. DB::ThreadStatus::initPerformanceCounters() @ 0xee5a96b in /usr/bin/clickhouse
7. DB::ThreadStatus::setupState(std::__1::shared_ptr<DB::ThreadGroupStatus> const&) @ 0xee5a648 in /usr/bin/clickhouse
8. DB::CurrentThread::initializeQuery() @ 0xee5c602 in /usr/bin/clickhouse
9. DB::BackgroundSchedulePool::attachToThreadGroup() @ 0xe85bfcd in /usr/bin/clickhouse
10. DB::BackgroundSchedulePool::threadFunction() @ 0xe85c0b2 in /usr/bin/clickhouse
11. ? @ 0xe85ceb2 in /usr/bin/clickhouse
12. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x84f700f in /usr/bin/clickhouse
13. ? @ 0x84faaa3 in /usr/bin/clickhouse
14. start_thread @ 0x74a4 in /lib/x86_64-linux-gnu/libpthread-2.24.so
15. clone @ 0xe8d0f in /lib/x86_64-linux-gnu/libc-2.24.so
 (version 21.2.5.5 (official build))
```

**Additional context**
CH runs in LXC container on host Proxmox VE  6.3-3 with 512GB RAM. Container has limit 128GB RAM.
Here are some settings from the config file:
```
<max_server_memory_usage>125000000000</max_server_memory_usage>
<max_server_memory_usage_to_ram_ratio>0.9</max_server_memory_usage_to_ram_ratio>
<total_memory_profiler_step>4194304</total_memory_profiler_step>
```
Here is output of "free" command:
```
# free
              total        used        free      shared  buff/cache   available
Mem:      134217728      396256   132932412        9376      889060   133821472
Swap:       8388604           0     8388604
```

**Comment by Maximus43** [comment]
> Do you use some memory limitation on LXC/cgroup level?
Yes, container has limit 131072 Mbytes for RAM

> Do you check `free` in container? Or outside?
It was inside the container.

Here is "free" on host machine (outside):

`              total        used        free      shared  buff/cache   available
Mem:      528327160    13392500   342633056     5944252   172301604   505513600
Swap:       8388604           0     8388604`

Host is server with ECC memory.

**Comment by Maximus43** [comment]
I'm slightly confused with startup message:
```
2021.03.10 15:17:19.060659 [ 1272 ] {} <Information> Application: Available RAM: 503.85 GiB; physical cores: 12; logical cores: 24.
2021.03.10 15:17:19.060687 [ 1272 ] {} <Information> Application: Ready for connections.
```

It shows amount of RAM on the host machine, while container is limited to 128GB of RAM.
Can it be a source of issue? Probably Clickhouse determines available RAM in LXC incorrectly. Or LXC virtualisation has not enough isolation from host. :-)

**Comment by filimonov** [comment]
It maybe related somehow. See
#16918
#2261
#18448
#17556

**Comment by Maximus43** [comment]
Yes, It seems that segfault is related to the wrong definition of available memory. "Bare-metal" and KVM instances have no issues at all.

**Comment by sergey-gt-ekb** [comment]
On official docker images since versions ~20, got the same crushes.... :(

CH totally not work in containers?!

Version 18.16 work more stable, without crushes. But i have got another issue(and same on latest):
At random moment server stop working normally, and in the server logs begin writes a lot of binary data such this.
```2021.05.08 23:34:11.237810 [ 41 ] {} <Error> ServerErrorHandler: Poco::Exception. Code: 1000, e.code() = 104, e.displayText() = Connection reset by peer (version 19.
2021.05.08 23:34:13.617173 [ 41 ] {} <Error> ServerErrorHandler: Poco::Exception. Code: 1000;QXèXFWXXfXX&?UXVXX{DXX@X<7ڱXXeXYtX<82>XX-T@G?qTvXXXXrUXXVXHX]XmXXtXXLPXE
X[|T9hXCXה8X=@XXX;RSBX]t_OXBF^\(X:@,K|=XX@R{X'htXPSlNXXBXXTdX[EXFX@EXC@XDBQ&DX|@XCXQeDCXN@XOX4XX!KX1_X@Xӄ7PXCXkDVpXX"HXCBXCXXDAXGX9XXF*5EX[de_XXjXX`:XXFX4KFXAXX^tVXY
X\5UX@XXX=XXXX4nXXX_XX.XXA,XT@/XAXZ<MS_XNwX3XBCdXt]L=AO´5N^X#zX ^XXXX<4LpXBXL08C@XSX<?CX7_t!XQCtgXEXX]XXUрvX2XXVcXX`6X<XZCXCXG*XAsBX[ޘCXTX3XX<7FC>}XQ.XXzKX7XS{XX6XXi
XA&XL5_XC^XLDXdAXX^BUXxX4ELLXE9LAX_AXXXNX=AZ*X&NXXUlT)FXXSNX=NDXpmXQX9PXU40A~a,GXqûUDL(H@ƂXHXgX9UT      XKX^l$6XX@XX<FXEXLX<4SbCXXXB]X$$\[XXXvX,=XXX^X\Z-XX)AX^l\XSUX
F4N`vB0XL-X4XXXtGXD<85>-X,?-@PX4Xé]l3X1XXXl]X)Xٳ<%vvXGXLMXXCZXX>X8ƈm,^T#CX)LfXQX;XXPF@@XA$!;tXiXX=XXXNBX3X7XX-LOXtAXbdKXXBXXt5fX@|$TK\XXRX<[RX@XX<      UXXX2XXXr]D@d
```

**Comment by alexey-milovidov** [comment]
@sergey-gt-ekb You should never use old versions of ClickHouse.
The only supported versions are 21.4, 21.3 and 21.2.

**Comment by alexey-milovidov** [comment]
@azat Could it be related to jemalloc?

**Comment by azat** [comment]
>@azat Could it be related to jemalloc?

@alexey-milovidov issue with 18.16 may be related to jemalloc, but NULL pointer in - I doubt.

**Comment by azat** [comment]
@Maximus43 do you have an ability to check some **development** version of clickhouse (I wouldn't recommend to do this on production data)? If so I would recommend you to check version with ASAN, for this you need to:
- [go to the commits list](https://github.com/ClickHouse/ClickHouse/commits/master)
- [press on checks report and go to `ClickHouse build check`](https://clickhouse-builds.s3.yandex.net/23970/02fd1d4bc3096d2b7e27e2edc94852098a847deb/clickhouse_build_check/report.html)
- get packages for the row with "address" in the "sanitizer" column

**Comment by azat** [comment]
BTW can you also provide:
- linux kernel version
- linux kernel config
- `sysctl -a`
- `/proc/meminfo`

**Comment by azat** [comment]
>It shows amount of RAM on the host machine, while container is limited to 128GB of RAM.

Yes, because clickhouse does not cgroup aware right now.

>Yes, It seems that segfault is related to the wrong definition of available memory. "Bare-metal" and KVM instances have no issues at all.

It should not be related to this, since you should get std::bad_alloc exception in case of memory allocation failure anyway

Plus you have `max_server_memory_usage` in the config, so clickhouse should use only this memory (and just a note if you set `max_server_memory_usage` you don't need `max_server_memory_usage_to_ram_ratio`)

**Comment by Maximus43** [comment]
> BTW can you also provide:
> 
> * linux kernel version
> * linux kernel config
> * `sysctl -a`
> * `/proc/meminfo`

I have migrated Clickhouse VM to full virtualisation (KVM), it works perfect.
Previously Clickhouse was installed on the following instance:
`Linux 5.4.78-2-pve #1 SMP PVE 5.4.78-2 (Thu, 03 Dec 2020 14:26:17 +0100) x86_64 GNU/Linux`

```
# cat /proc/meminfo
MemTotal:       528327452 kB
MemFree:        181880448 kB
MemAvailable:   449090992 kB
Buffers:         4247652 kB
Cached:         277240836 kB
SwapCached:         4400 kB
Active:         195762620 kB
Inactive:       138077592 kB
Active(anon):   67389540 kB
Inactive(anon):  3479928 kB
Active(file):   128373080 kB
Inactive(file): 134597664 kB
Unevictable:      159972 kB
Mlocked:          159972 kB
SwapTotal:       8388604 kB
SwapFree:        8365060 kB
Dirty:              2508 kB
Writeback:             0 kB
AnonPages:      52506856 kB
Mapped:         18498100 kB
Shmem:          18525216 kB
KReclaimable:    7661924 kB
Slab:            8771028 kB
SReclaimable:    7661924 kB
SUnreclaim:      1109104 kB
KernelStack:       26080 kB
PageTables:      1299808 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    272552328 kB
Committed_AS:   234206716 kB
VmallocTotal:   34359738367 kB
VmallocUsed:     1699292 kB
VmallocChunk:          0 kB
Percpu:            69376 kB
HardwareCorrupted:     0 kB
AnonHugePages:  31889408 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:     3075524 kB
DirectMap2M:    336607232 kB
DirectMap1G:    197132288 kB
```
[sysctl.txt](https://github.com/ClickHouse/ClickHouse/files/6451505/sysctl.txt)

[kernel_config.txt](https://github.com/ClickHouse/ClickHouse/files/6451507/kernel_config.txt)

**Comment by azat** [comment]
@Maximus43 

>Linux 5.4.78-2-pve #1 SMP PVE 5.4.78-2 (Thu, 03 Dec 2020 14:26:17 +0100) x86_64 GNU/Linux

Can you try latest pve kernel? (5.11)
And also can you try disabling huge pages?

@sergey-gt-ekb 

>kernel 3.10.0-1062.4.2.vz7.116.7

Too old kernel (first release of 3.10 kernel was in 2013), can you please try latest kernel?

**Comment by alexey-milovidov** [comment]
No info provided.

**Comment by alexey-milovidov** [comment]
Fixed in #32916.
